File: RecoverBinarySearchTree.java
/**
 * Created by yangchi on 10/15/14.
 */
public class RecoverBST {

    public TreeNode first = null;
    public TreeNode second = null;
    public TreeNode pre = null;

    public void recoverTree(TreeNode root) {
        if (root == null) {
            return;
        }
        //inOrder(root, first, second, pre);
        inOrder(root);
        if(first == null)
            System.out.println("First is empty");
        if (second == null)
            System.out.println("Second is empty");
        if(first == null || second == null)
            return;
        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }

    public void inOrder(TreeNode root) { //, TreeNode first, TreeNode second, TreeNode pre) {
        if(root == null) {
            return;
        }
        inOrder(root.left); //, first, second, pre);
        if(pre == null) {
            pre = root;
        } else {
            if(pre.val > root.val) {
                if(first == null) {
                    first = pre;
                    System.out.println("Changed first to " + pre.val);
                }
                second = root;
                System.out.println("Changed second to " + root.val);
            }
            pre = root;
        }
        inOrder(root.right); //, first, second, pre);
    }

    public static void main(String args[]) {
        TreeNode root = new TreeNode(0);
        root.left = new TreeNode(1);
        RecoverBST bst = new RecoverBST();
        bst.recoverTree(root);

    }

}

File: RecoverBinarySearchTree.cc
#include <iostream>


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    void recoverTree(TreeNode *root) {
        TreeNode *pre = NULL, *first = NULL, *second = NULL;
        recoverTree(root, &pre, &first, &second);
        int temp = first->val;
        first->val = second->val;
        second->val = temp;
        return;
    }

    void recoverTree(TreeNode *node, TreeNode **pre, TreeNode **first, TreeNode **second) {
        if (!node) {
            return;
        }

        recoverTree(node->left, pre, first, second);
        if (pre && *pre && (*pre)->val > node->val) {
            if (!*first) {
                *first = *pre;
            }
            *second = node;
        } else {
            *pre = node;
        }

        recoverTree(node->right, pre, first, second);
    }
};

int main()
{
    Solution sol;
    TreeNode *root = new TreeNode(0);
    root->left = new TreeNode(1);
    sol.recoverTree(root);
}
